libevmasm/ThirdOptimizerTesting.cpp

```diff
@@ -0,0 +1,368 @@
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libevmasm/ThirdOptimizerTesting.h>
+#include "ThirdOptimizerTesting.h"
+#include <algorithm>
+#include <tuple>
+
+template<typename It, typename = typename std::enable_if<is_assembly_item<decltype(*std::declval<It>())>::value>::type>
+It noReturnValueBegin(It _begin, It _end);
+
+template<typename It, typename = typename std::enable_if<is_assembly_item<decltype(*std::declval<It>())>::value>::type>
+It backExpressionBegin(It _begin, It _end, bool _allowNoReturn = false)
+{
+	AssemblyItem const& topLevelItem = *(_end - 1);
+
+	signed arguments = topLevelItem.arguments();
+	auto currentEnd = _end - 1 /* top-level operation */;
+	while (arguments > 0)
+	{
+		// TODO this code might break if anything returns more than 1 value, since we can't split things up
+		currentEnd = noReturnValueBegin(_begin, currentEnd);
+		arguments -= static_cast<AssemblyItem const&>(*(currentEnd - 1)).returnValues();
+		currentEnd = backExpressionBegin(_begin, currentEnd, false);
+		assertThrow(currentEnd >= _begin, OptimizerException, "currentEnd is before begin");
+	}
+
+	if (!_allowNoReturn && topLevelItem.returnValues() == 0) return backExpressionBegin(_begin, currentEnd, false);
+
+	return currentEnd;
+}
+
+template<typename It, typename = typename std::enable_if<is_assembly_item<decltype(*std::declval<It>())>::value>::type>
+It noReturnValueBegin(It _begin, It _end)
+{
+	assertThrow(_begin != _end, OptimizerException, "expression is empty");
+	AssemblyItem const& topLevelItem = *(_end - 1);
+	if (topLevelItem.returnValues() != 0) return _end;
+	return backExpressionBegin(_begin, _end, true);
+}
+
+template<typename It, typename = typename std::enable_if<is_assembly_item<decltype(*std::declval<It>())>::value>::type>
+vector<It> parseExpressions(It _begin, It _end, signed _limit = -1, bool _allowNoReturn = false)
+{
+	auto currentEnd = _end;
+	std::vector<It> expressions{};
+	while (currentEnd != _begin && (_limit == -1 || signed(expressions.size()) < _limit))
+	{
+		auto oldEnd = currentEnd;
+		currentEnd = backExpressionBegin(_begin, oldEnd, _allowNoReturn);
+		expressions.push_back(currentEnd);
+		assertThrow(currentEnd >= _begin, OptimizerException, "currentEnd is before begin");
+	}
+
+	return expressions;
+}
+
+template<typename It, typename = typename std::enable_if<is_assembly_item<decltype(*std::declval<It>())>::value>::type>
```

Comment:
Is the purpose of this to work both for `const` and `non-const`?

Reply:
Well, the purpose of this line specifically is to ensure that no programmer tries to pass in some random iterator. `is_assembly_item` is defined such that it can be any const/non-const iterator.